packages/gossip/stream.go

```diff
@@ -0,0 +1,183 @@
+
+	if conf.useDefaultTimeout {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, defaultConnectionTimeout)
+		defer cancel()
+	}
+
+	stream, err := m.libp2pHost.NewStream(ctx, libp2pID, protocolID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "dial %s / %s failed", address, p.ID())
+	}
+	// Disable libp2p lazy negotiation.
+	if _, err := stream.Write(nil); err != nil {
+		return nil, errors.Wrapf(err, "dial %s / %s failed", address, p.ID())
+	}
+	m.log.Debugw("outgoing connection established",
+		"id", p.ID(),
+		"addr", stream.Conn().RemoteMultiaddr(),
+	)
+	return stream, nil
+}
+
+func (m *Manager) acceptPeer(ctx context.Context, p *peer.Peer, opts []ConnectPeerOption) (network.Stream, error) {
+	gossipEndpoint := p.Services().Get(service.GossipKey)
+	if gossipEndpoint == nil {
+		return nil, ErrNoGossip
+	}
+
+	stream, err := func() (network.Stream, error) {
+		// wait for the connection
+		m.acceptWG.Add(1)
+		defer m.acceptWG.Done()
+		conf := buildConnectPeerConfig(opts)
+		if conf.useDefaultTimeout {
+			var cancel context.CancelFunc
+			ctx, cancel = context.WithTimeout(ctx, defaultConnectionTimeout)
+			defer cancel()
+		}
+		am, err := newAcceptMatcher(p)
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+		if ok := m.setAcceptMatcher(am); !ok {
+			return nil, errors.WithStack(ErrDuplicateAccept)
+		}
+		defer m.removeAcceptMatcher(am)
+		select {
+		case stream := <-am.streamCh:
+			return stream, nil
+		case <-ctx.Done():
+			err := ctx.Err()
+			if errors.Is(err, context.DeadlineExceeded) {
+				m.log.Debugw("accept timeout", "id", am.peer.ID())
+				return nil, errors.WithStack(ErrTimeout)
+			}
+			m.log.Debugw("context error", "id", am.peer.ID(), "err", err)
+			return nil, errors.WithStack(err)
+		}
+	}()
+	if err != nil {
+		return nil, fmt.Errorf(
+			"accept %s / %s failed: %w",
+			net.JoinHostPort(p.IP().String(), strconv.Itoa(gossipEndpoint.Port())), p.ID(),
+			err,
+		)
+	}
+	m.log.Debugw("incoming connection established",
+		"id", p.ID(),
+		"addr", stream.Conn().RemoteMultiaddr(),
+	)
+	return stream, nil
+}
+
+type acceptMatcher struct {
+	peer     *peer.Peer // connecting peer
+	libp2pID libp2ppeer.ID
+	streamCh chan network.Stream
+}
+
+func newAcceptMatcher(p *peer.Peer) (*acceptMatcher, error) {
+	libp2pID, err := libp2putil.ToLibp2pPeerID(p)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	return &acceptMatcher{
+		peer:     p,
+		libp2pID: libp2pID,
+		streamCh: make(chan network.Stream, 1),
+	}, nil
+}
+
+func (m *Manager) setAcceptMatcher(am *acceptMatcher) bool {
+	m.acceptMutex.Lock()
+	defer m.acceptMutex.Unlock()
+	_, exists := m.acceptMap[am.libp2pID]
+	if exists {
+		return false
+	}
+	m.acceptMap[am.libp2pID] = am
+	return true
+}
+
+func (m *Manager) removeAcceptMatcher(am *acceptMatcher) {
+	m.acceptMutex.Lock()
+	defer m.acceptMutex.Unlock()
+	delete(m.acceptMap, am.libp2pID)
+}
+
+func (m *Manager) streamHandler(stream network.Stream) {
```

Comment:
How/When exactly is this called? `m.libp2pHost.SetStreamHandler(protocolID, m.streamHandler)` in `NewManager` sets this but I don't understand how it is called while we set up the connection because that is the only time the `acceptMatcher` exists if I understand correctly. (see `defer m.removeAcceptMatcher(am)` in `acceptPeer`)

Reply:
After we register `streamHandler` with `. SetStreamHandler()`, libp2p will call our handler function every time there is a new incoming connection (stream). 
In the handler function, we check are we accepting this connection or not, if yes we match it and set up a new neighbor.